File: 1주차/kmdngyu/HappyNumber.java
package com.kmdgnyu;

import java.util.ArrayList;
import java.util.List;

public class HappyNumber {
    public static boolean isHappy(int input) {
        List<Integer> cache = new ArrayList<>();
        return getIsHappy(input, cache);
    }

    private static boolean getIsHappy(int input, List<Integer> cache) {
        if (input == 1) {
            return true;
        }
        List<Integer> numbers = new ArrayList<>();
        int maxDigit = 10;
        for (int i = maxDigit; i >= 0; i--) {
            int digit = (int) Math.pow(10, i);
            int quotient = input / digit;
            if (quotient > 0) {
                numbers.add(quotient);
                input = input - quotient * digit;
            }
        }
        int sum = 0;
        for (Integer number : numbers) {
            sum += number * number;
        }
        if (cache.contains(sum)) {
            return false;
        } else {
            cache.add(sum);
        }
        return getIsHappy(sum, cache);
    }
}

File: 1주차/simsimjae/HappyNumber.js
var isHappy = function(n) {
    const MAP = new Map();

    const squareSum = function(number){
        let sum = 0;
        while(number > 0) {
            const remain = parseInt(number % 10, 10);
            sum += remain * remain;
            number = parseInt(number / 10, 10);
        } 
        return sum;
    };
    
    while(true){
        const sum = squareSum(n);
        console.log(sum);
        if(MAP.has(sum)) return false;
        else MAP.set(sum);
        
        if(sum === 1) return true;
        else n = sum;
    }
};
